File: app/src/main/java/com/ndmsystems/coala/helpers/Hex.java
package com.ndmsystems.coala.helpers;


/**
 * Created by bas on 25.10.16.
 */
public class Hex {
    private static final char[] DIGITS_LOWER;
    private static final char[] DIGITS_UPPER;

    public static byte[] decodeHex(char[] data){
        int len = data.length;
            byte[] out = new byte[len >> 1];
            int i = 0;

            for(int j = 0; j < len; ++i) {
                int f = toDigit(data[j], j) << 4;
                ++j;
                f |= toDigit(data[j], j);
                ++j;
                out[i] = (byte)(f & 255);
            }

            return out;
    }

    protected static int toDigit(char ch, int index) {
        return Character.digit(ch, 16);
    }

    public static char[] encodeHex(byte[] data) {
        return encodeHex(data, true);
    }

    public static char[] encodeHex(byte[] data, boolean toLowerCase) {
        return encodeHex(data, toLowerCase?DIGITS_LOWER:DIGITS_UPPER);
    }

    protected static char[] encodeHex(byte[] data, char[] toDigits) {
        int l = data.length;
        char[] out = new char[l << 1];
        int i = 0;

        for(int j = 0; i < l; ++i) {
            out[j++] = toDigits[(240 & data[i]) >>> 4];
            out[j++] = toDigits[15 & data[i]];
        }

        return out;
    }

    public static String encodeHexString(byte[] data) {
        if (data == null) return "";
        return new String(encodeHex(data));
    }

    static {
        DIGITS_LOWER = new char[]{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
        DIGITS_UPPER = new char[]{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
    }
}

File: app/src/main/java/com/ndmsystems/coala/helpers/Hex.kt
package com.ndmsystems.coala.helpers

/**
 * Created by bas on 25.10.16.
 */
object Hex {
    private val DIGITS_LOWER: CharArray
    private val DIGITS_UPPER: CharArray
    @JvmStatic
    fun decodeHex(data: CharArray): ByteArray {
        val len = data.size
        val out = ByteArray(len shr 1)
        var i = 0
        var j = 0
        while (j < len) {
            var f = toDigit(data[j], j) shl 4
            ++j
            f = f or toDigit(data[j], j)
            ++j
            out[i] = (f and 255).toByte()
            ++i
        }
        return out
    }

    internal fun toDigit(ch: Char, index: Int): Int {
        return ch.digitToIntOrNull(16) ?: -1
    }

    @JvmOverloads
    fun encodeHex(data: ByteArray, toLowerCase: Boolean = true): CharArray {
        return encodeHex(data, if (toLowerCase) DIGITS_LOWER else DIGITS_UPPER)
    }

    internal fun encodeHex(data: ByteArray, toDigits: CharArray): CharArray {
        val l = data.size
        val out = CharArray(l shl 1)
        var i = 0
        var j = 0
        while (i < l) {
            out[j++] = toDigits[240 and data[i].toInt() ushr 4]
            out[j++] = toDigits[15 and data[i].toInt()]
            ++i
        }
        return out
    }

    @JvmStatic
    fun encodeHexString(data: ByteArray?): String {
        return if (data == null) "" else String(encodeHex(data))
    }

    init {
        DIGITS_LOWER = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
        DIGITS_UPPER = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
    }
}